Paintroid/src/main/java/org/catrobat/paintroid/ui/Perspective.kt

```diff
@@ -0,0 +1,196 @@
+/*
+ * Paintroid: An image manipulation application for Android.
+ * Copyright (C) 2010-2021 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.paintroid.ui
+
+import android.graphics.Canvas
+import android.graphics.PointF
+import android.graphics.Rect
+import androidx.annotation.VisibleForTesting
+import kotlin.jvm.Synchronized
+import kotlin.math.max
+import kotlin.math.min
+
+const val MIN_SCALE = 0.1f
+const val MAX_SCALE = 100f
+private const val SCROLL_BORDER = 50f
+private const val BORDER_ZOOM_FACTOR = 0.95f
+
+open class Perspective(bitmapWidth: Int, bitmapHeight: Int) {
+    @VisibleForTesting
+    @JvmField
+    var surfaceWidth = 0
+
+    @VisibleForTesting
+    @JvmField
+    var surfaceHeight = 0
+
+    @VisibleForTesting
+    @JvmField
+    var surfaceCenterX = 0f
+
+    @VisibleForTesting
+    @JvmField
+    var surfaceCenterY = 0f
+
+    @VisibleForTesting
+    var surfaceScale = 1f
+
+    @VisibleForTesting
+    @JvmField
+    var surfaceTranslationX = 0f
+
+    @VisibleForTesting
+    @JvmField
+    var surfaceTranslationY = 0f
+    var fullscreen = false
+
+    @VisibleForTesting
+    var initialTranslationY = 0f
+
+    @set:Synchronized
+    var scale: Float
+        get() = surfaceScale
+        set(scale) {
+            surfaceScale = max(MIN_SCALE, min(MAX_SCALE, scale))
+        }
+
+    val scaleForCenterBitmap: Float
+        get() {
+            val screenSizeRatio = surfaceWidth.toFloat() / surfaceHeight
+            val bitmapSizeRatio = bitmapWidth / bitmapHeight
+            var ratioDependentScale = if (screenSizeRatio > bitmapSizeRatio) {
+                surfaceHeight / bitmapHeight
+            } else {
+                surfaceWidth / bitmapWidth
+            }
+            ratioDependentScale = min(ratioDependentScale, 1f)
+            ratioDependentScale = max(ratioDependentScale, MIN_SCALE)
+            return ratioDependentScale
+        }
+
+    private var bitmapWidth: Float
+    private var bitmapHeight: Float
+    private var initialTranslationX = 0f
+
+    init {
+        this.bitmapWidth = bitmapWidth.toFloat()
```

Comment:
but then we can also remove this init block and other occurences where we cast `bitmapWidth` and `bitmapHeight` to float